### ISSUE TITLE
SRI hash for ipaddr.js does not match

### ISSUE THREAD

**Issue by Tugzrida** [cdnjs/cdnjs#14080]
I'm seeing some quite odd behaviour with the [ipaddr.js](https://cdnjs.com/libraries/ipaddr.js) library:

Firstly, here is the correct SRI hash for ipaddr.min.js@2.0.1. I've crosschecked this between the original file in the source project, the file served by cdnjs, and the same file served by another CDN and they all match.

    sha512-HLkOIPULXMWy1Z/SOoR5PWofAu7iM5OwJqMMkCiqGR+wijlXbRj+D3jeBTK6XjyO43vWvBVgQyFyhCDi+Pztdg==

When I go to the [library's page on the cdnjs website](https://cdnjs.com/libraries/ipaddr.js), the "Copy SRI Hash" and "Copy Script Tag" buttons both provide this hash:

    sha512-GcJ0uemx9BQwsZusgXAGzw6/BDAkqt0EUbtq8QAtTuDlAa3AVQ0zb+Zg9xqY4yXP47CHBv2pO1V0D78gtWYNgw==

I have also just confirmed this behaviour via a VPN to the UK, so it doesn't seem to just be a cached invalid value in my local CF edge.

**Comment by MattIPv4** [comment]
logs: https://github.com/cdnjs/logs/tree/prod/packages/i/ipaddr.js/2.0.1

**Comment by Tugzrida** [comment]
If it makes any difference, I knew the new version was coming, so I did have the library page open and reloaded it a few times waiting for it to make it through the pipeline, but given that those logs have the same incorrect hash, I'm guessing the web frontend probably had nothing to do with it

**Comment by MattIPv4** [comment]
Yeah, it definitely looks like we've somehow got the wrong hash calculated for it. I've confirmed observing the same as you, both the incorrect hash and what the correct hash should be.

It looks like it attempts to compress and calculate the hash for the file twice, so I wonder if that's the issue (before it then also appears to have paniced):

https://github.com/cdnjs/logs/blob/7a17e3c8525ae2a899cd09e6bc1170cc11c32a72/packages/i/ipaddr.js/2.0.1/processing.log#L6-L7

https://github.com/cdnjs/logs/blob/7a17e3c8525ae2a899cd09e6bc1170cc11c32a72/packages/i/ipaddr.js/2.0.1/processing.log#L12-L14

**Comment by Tugzrida** [comment]
I'm not quite sure what the file tree looks like on your worker environment here, but the first SRI generation was for `/tmp/work/ipaddr.min.js`, and then the second was for `/tmp/work/lib/ipaddr.min.js` (note the `lib` folder)

I'm guessing something got confused by the project providing both an unminified and pre-minified file, then minified the unminified file, hashed it, but then published the pre-minified version?

**Comment by MattIPv4** [comment]
Looping in https://github.com/cdnjs/cdnjs/discussions/14124 as it seems to be very related to this

**Comment by Tugzrida** [comment]
Ok, I've spent quite a few hours poring over [cdnjs/tools](https://github.com/cdnjs/tools), and the logs for both [ipaddr.js](https://github.com/cdnjs/logs/blob/prod/packages/i/ipaddr.js/2.0.1/processing.log) and [moment.js](https://github.com/cdnjs/logs/blob/prod/packages/m/moment.js/2.29.2/processing.log) (as mentioned in #14124) to try and get to the bottom of this.

Part of the cause is contained in [this snippet](https://github.com/cdnjs/tools/blob/e87c098c9dc2dce6b3f49c77adf4f29b60eb5366/packages/packages.go#L200-L203):

```go
out = append(out, NpmFileMoveOp{
	From: path.Join(*fileMap.BasePath, f),
	To:   f,
})
```

`f` in this case is a match from a glob in the package's config JSON.  You can see that `To`, which determines the file's destination in cdnjs, does not include the `basePath` option from the config like `From` does. This is most likely intended behaviour as it removes unnecessary folder hierarchy from the CDN filesystem.

It however leads to issues when there are identically-named files in different folders of the library, or files that become identically-named once minified.

As an example, here's ipaddr.js' config `fileMap`:
```json
"fileMap": [
  {
    "basePath": "",
    "files": [
      "ipaddr*"
    ]
  },
  {
    "basePath": "lib",
    "files": [
      "ipaddr*"
    ]
  }
]
```

And here's the files which that config selects from the npm package:

```
node_modules/ipaddr.js/
├── ipaddr.min.js
└── lib
    ├── ipaddr.js
    └── ipaddr.js.d.ts
```

The next thing the script does is minify `lib/ipaddr.js` (in the [`optimizeWorker`](https://github.com/cdnjs/tools/blob/e87c098c9dc2dce6b3f49c77adf4f29b60eb5366/cmd/process-version/main.go#L202-L236) function). In the [logs](https://github.com/cdnjs/logs/blob/prod/packages/i/ipaddr.js/2.0.1/processing.log#L4), you can see that `ipaddr.min.js` is detected as already minified and so isn't re-minified.

The directory now looks like this, and here we see the problem emerge:

```
node_modules/ipaddr.js/
├── ipaddr.min.js
└── lib
    ├── ipaddr.js
    ├── ipaddr.min.js
    └── ipaddr.js.d.ts
```

When brotli/gzip compressed, SRI hashed and copied to the output directory (in the [`emitFromWorkspace`](https://github.com/cdnjs/tools/blob/e87c098c9dc2dce6b3f49c77adf4f29b60eb5366/cmd/process-version/main.go#L83-L110) function), the `lib` folder is dropped as it's part of `basePath`, not the glob. We should thus end up with an output directory looking like this (with brotli copies of the files hidden for clarity):

```
output
├── ipaddr.min.js.gz
├── ipaddr.min.js.sri
|
├── ipaddr.js.gz
├── ipaddr.js.sri
|
├── ipaddr.min.js.gz
├── ipaddr.min.js.sri
|
├── ipaddr.js.d.ts.gz
└── ipaddr.js.d.ts.sri
```

You can however see that `ipaddr.min.js` is duplicated, and this is the crux of the issue.

In practice, not all of these files were actually written to the output directory, because the brotli command won't overwrite files without the `-f` option. This caused the crash shown in the [log](https://github.com/cdnjs/logs/blob/efb23af632d8e0e7d165d97422884bbaf689ddd3/packages/i/ipaddr.js/2.0.1/processing.log#L14).

Even if brotli was run with `-f`, the script would not be guaranteed to output a matching SRI and file content (or even matching file content between the brotli and gzip versions of the file) because the script is multithreaded.

---

## Root cause

So, the root cause is that minification can create duplicately-named files in different directories, if both a non-minified and pre-minified version of a file are selected from a source package, and these two versions are in different directories. This is because the [JS](https://github.com/cdnjs/tools/blob/e87c098c9dc2dce6b3f49c77adf4f29b60eb5366/compress/js.go#L23-L27) and [CSS](https://github.com/cdnjs/tools/blob/e87c098c9dc2dce6b3f49c77adf4f29b60eb5366/compress/css.go#L22-L26) minifier code only checks for an already minified version of the file in the current directory.

The same issue could occur if a source package already includes duplicately-named files in different directories.

## Possible solutions

1. Modify the [section of `NpmFilesFrom`](https://github.com/cdnjs/tools/blob/e87c098c9dc2dce6b3f49c77adf4f29b60eb5366/packages/packages.go#L200-L203) I show in my first snippet, so that `To` also contains the config's basePath. As I mentioned, this would make the CDN's paths contain unnecessary folders from the source library, but it would solve every variant of the issue. It would however also result in multiple minified copies of some files existing in the CDN.

2. Add requirements that package JSON definitions don't select duplicate files, possibly in the CI for pull requests. I feel that this is the neatest option, but would not catch faulty configs that already exist like moment.js and ipaddr.js. Additionally, if the structure of the source package changes, duplicates could come into existence at a later stage.

    In practice, this means only selecting pre-minified versions of files if they exist in the same folder as the non-minified file. This way, the variants will be detected correctly and used as-is from the package. For libraries like moment and ipaddr where the minified files are in a different folder to the unminified source, it would mean selecting only the unminified file in the package's JSON file, and then minifying the file in the processing pipeline rather than using the pre-minified copy.

3. Improve how the JS and CSS minifiers look for pre-minified in the source package. This could be accomplished by copying the source files to an intermediate output directory prior to brotli/gz compression and SRI hashing, and prior to starting the multithreaded section of the build script. This way, duplicates would be overwritten completely if they did exist, and pre-minified files would be detected correctly by the existing minification logic. I think this is the best all-round solution.

**Comment by MattIPv4** [comment]
👀 Ty for digging into this!

I don't particularly feel that options 1 or 2 are overly viable. Choosing option 1 would essentially make the `basePath` component of our file maps redundant, and option 2 doesn't seem maintainable long-term as we'd have to continually fix existing packages if version updates cause clashes.

I think 3 is probably on the right tracks, but from what you've said, I'm wondering if the minification is even the cause, or rather just a complicating factor to the pure race condition of processing two assets?

For example, if the following config was used:

```json
[
  {
    "basePath": "",
    "files": [ "*.js" ]
  },
  {
    "basePath": "test",
    "files": [ "*.js" ]
  }
]
```

With these files:

```text
package
├── a.js
└── test
    └── a.js
```

Would there not be the same risk of the race condition for SRI processing there too?

To me, it seems like the solution would be just to make sure that the processing order of files is deterministic always, so that we avoid the race condition for SRI processing. Or, as you suggested, split the package destructing/minification step away from the SRI processing (so we generate the final file structure for the CDN first, and then calculate the SRI hashes after)?

**Comment by dregad** [comment]
I hope you guys can fix this issue, which is affecting us (@mantisbt) since yesterday, as we had to update moment.js to 2.29.2 to address CVE-2022-24785 (see https://github.com/advisories/GHSA-8hfj-j24r-96c4).

It is not acceptable for us to revert to an earlier version of moment.js due to the security issue, so it would be awesome if you guys could manually fix the SRI quickly, as a stopgap measure until a permanent fix can be found for the root cause.

Thanks in advance !

**Comment by Tugzrida** [comment]
Thanks for that point @dregad. That should be able to happen, but I'm not sure how quickly it could be done as I don't know the details of the Cloudflare processes behind cdnjs.

As a workaround in the meantime, you could calculate the hash of the file yourself (from memory openssl can do it fairly easily), or use the non-minified file and hash from cdnjs(you'll have to click the "show all files" button), as it isn't effected by this bug.

Currently I'm working on narrowing down the behaviour of the bug so @MattIPv4 and I can find the best permanent solution.

**Comment by dregad** [comment]
> calculate the hash of the file yourself

Thanks for the tip @Tugzrida ; I just tried to that but I'm getting a 404 on https://cdnjs.cloudflare.com/ajax/libs/moment.js/2.29.2/moment-with-locales.min.js :confused: am I doing something wrong, or is there some other problem ?

**Comment by Tugzrida** [comment]
Ah darn, I had forgotten that that happened. Because the processing script crashed partway through due to the duplicate files bug, some of the files ended up empty.

If you need it fixed fast, I'd just suggest temporarily hosting the file yourself or using a different CDN, as I'm not sure how long it'll take to get this fixed.

**Comment by dregad** [comment]
OK, thanks for the feedback ! I thought I was going nuts for a while, since I was seeing a 200 in the browser console, but a 404 with wget... Very confusing.

![image](https://user-images.githubusercontent.com/449891/163351063-d42a26a6-8112-4dbd-9420-75d899b03897.png)

```
$ wget https://cdnjs.cloudflare.com/ajax/libs/moment.js/2.29.2/moment-with-locales.min.js -nv
https://cdnjs.cloudflare.com/ajax/libs/moment.js/2.29.2/moment-with-locales.min.js:
2022-04-14 10:56:42 ERROR 404: Not Found.
```

> temporarily hosting the file yourself or using a different CDN

MantisBT can be told not to use CDN (it's a simple setting to change in a config file), so it's probably I the best option for us, I guess we can live with that for a while.

Switching  to [jsdelivr](https://cdn.jsdelivr.net/npm/moment@2.29.2/min/moment-with-locales.min.js) is possible too, but more complex as it would require us to publish a new release. I would rather avoid it, unless we're looking at weeks to get this issue solved... 

Can you guys provide an ETA for a fix ?

**Comment by Tugzrida** [comment]
👋🏻 @MattIPv4

You're correct that the issue exists with same-named files independently from minification. That perhaps got lost in my long post 😅

To summarise the conditions for the bug to occur:

- The source package must include multiple files with the same name(ignoring `.min`) in different folders
- The glob(s) that selects these files for inclusion in cdnjs must produce the same path fragment for both files

And the effects:

- Each file processed by the script becomes three files: a brotli compressed version, a gzip compressed version, and the SRI hash file. Because the script is multi-threaded, when there are multiple input files mapping to these same three output files, a race condition exists, where in the worst case, all three outputted files derive from different input files, meaning users could be served different contents if their browser supports gzip or brotli, and the SRI hash may not match either version.
- As @dregad's experience reminded me, because the script currently crashes in these scenarios, empty files could also be written out when the crash occurs partway through writing the file.

Assuming we want to retain the multi-threaded behaviour of the processing script for performance reasons, I think it's most straightforward to add in a seperate destructuring step into an intermediate folder. By creating the structure of the CDN filesystem before the multi-threaded compressing, minifying and hashing process, the nature of filesystems not allowing duplicate names will ensure that duplicate files are resolved. Because the file mapping is 1:1 at this point, no race condition exists(and in the case of minification creating a 1:2 mapping, the logic of the minification scripts already looks for pre-minified versions).

You could also implement that behaviour purely in the script, but it would be more complicated than just using the nature of the filesystem to resolve duplicates.

I think it's also worth mentioning that because multiple `basePath` attributes are allowed per package, and because this attribute is not included in output filenames, it is possible that two files with the same name in a package could be selected, even if these files were truly different and not just minified and non-minified versions. Modifying the file structure of a package in a non-uniform way like this could always lead to one of the files missing from the output. I think this question is more for future consideration, as it doesn't impact the current problems with moment and ipaddr.

---

To sum up, going with option 3 and doing the package destructuring prior to starting multi-threaded operation would ensure that the gzip, brotli and SRI outputs for a given file would always correlate, however where identically named files exist in the source package(this time not including minified and non-minified versions of the same file), and they are selected by globs that return identical path fragments, only one of the files will make it to the output. That's because of the `basePath` issue I mention in the previous paragraph.

I'd love to help with a PR, but I have very little Golang experience, so it's not something I would be confident in. I am of course happy to continue helping with ideas and design.

**Comment by Tugzrida** [comment]
> OK, thanks for the feedback ! I thought I was going nuts for a while, since I was seeing a 200 in the browser console, but a 404 with wget... Very confusing.

Yes, I believe that'll be because the script crashed before creating the gzip version of that file. So when sending a request without an `Accept-Encoding` header (as I'm assuming wget would be doing), Cloudflare will be looking for the gzip version of the file to decompress, which doesn't exist, hence 404. The brotli version of the file however does exist (even though it's empty), so a browser with brotli support will get a 200, just with no content.

> Can you guys provide an ETA for a fix ?

That would be something @MattIPv4 would have to answer. Theoretically the correct files could be manually generated and loaded into Cloudflare, but again I don't know what the processes are here.

**Comment by MattIPv4** [comment]
Thanks for the further context. I agree that the simplest fix does seem to be to have the final file structure in an intermediate step before compression and SRI calcs.

I've escalated this to the folks at Cloudflare for attention, and will see if I can get an ETA from them!

**Comment by xtuc** [comment]
We deployed a fix and restarted the processing of ipaddr.js v2.0.1. Passed without any errors: https://github.com/cdnjs/logs/commit/4b01712ea9119fb49ee3433f40ae09d6980ace86.

Could you please check if SRIs are matching now?

**Comment by dregad** [comment]
Did you update moment.js as well? If so I can test and provide feedback tonight

**Comment by Tugzrida** [comment]
Thanks @xtuc

There's a few things that don't look quite right, going off the kv log at https://github.com/cdnjs/logs/commit/3f7f85c4509aee57b8fe828b3839267211091190

- The new file `ipaddr.js` has the correct SRI in that log. This file didn't previously exist as the script crashed before processing it
- Despite the SRI for `ipaddr.min.js` being changed in https://github.com/cdnjs/logs/commit/769c3cb1c635a606d1839d74790a82ef1bd31337, that hash is still not correct. The correct hash from the source package is `sha512-HLkOIPULXMWy1Z/SOoR5PWofAu7iM5OwJqMMkCiqGR+wijlXbRj+D3jeBTK6XjyO43vWvBVgQyFyhCDi+Pztdg==`
- I can't verify end-user functionality as the updated version is not visible in the cdnjs/cdnjs repo, nor cdnjs.com. The website still returns the wrong SRI hash.

**Comment by dregad** [comment]
Good morning,

I just tested and confirm that moment-with-locales.min.js 2.29.3 is working just fine now, so I'll upgrade to that.

It would still be good, if you guys could apply the fix to 2.29.2, which remains broken.

**Comment by CleyFaye** [comment]
Hi,

I think it is worth mentionning that although moment-with-locales.min.js works fine with version 2.29.3 it is not the case with moment.min.js.

The actual hash I get is different from the one provided on cdnjs.

```
$ curl -s https://cdnjs.cloudflare.com/ajax/libs/moment.js/2.29.3/moment.min.js | sha512sum | awk '{ print $1; }' | xxd -r -p | base64 -w 100
dGM81hdEjiW5IomkknOx3fIfwij3c7xwh6TcrbumlkHJtO81OKNjLISJaPEhCgVxM6+0uGJ7KRmI2YJFn0AmHQ==
```

Provided SRI hash: `sha512-x/vqovXY/Q4b+rNjgiheBsA/vbWA3IVvsS8lkQSX1gQ4ggSJx38oI2vREZXpTzhAv6tNUaX81E7QBBzkpDQayA==`

**Comment by Tugzrida** [comment]
Hi @MattIPv4 and @xtuc

Is there an update on a fix for this? The problem is still occurring with new versions of packages as mentioned by @kierenpitts in https://github.com/cdnjs/cdnjs/discussions/14124#discussioncomment-3261722

**Comment by MattIPv4** [comment]
I have asked internally again to see if I can get the correct :eyes: on this :)

**Comment by drees** [comment]
FWIW, moment.js 2.29.4 is broken.

`$ curl -s https://cdnjs.cloudflare.com/ajax/libs/moment.js/2.29.4/moment.min.js | sha512sum | awk '{ print $1; }' | xxd -r -p | base64 -w 100
+H4iLjY3JsKiF2V6N366in5IQHj2uEsGV7Pp/GRcm0fn76aPAk5V8xB6n8fQhhSonTqTXs/klFz4D0GIn6Br9g==`

The SRI hash on https://cdnjs.com/libraries/moment.js for 2.29.4 is:
`sha512-CryKbMe7sjSCDPl18jtJI5DR5jtkUWxPXWaLCst6QjH8wxDexfRJic2WRmRXmstr2Y8SxDDWuBO6CQC6IE4KTA==`

**Comment by MattIPv4** [comment]
I'll re-escalate this again 👍

**Comment by stijnb1234** [comment]
Same for moment-timezone

```
<script src="https://cdnjs.cloudflare.com/ajax/libs/moment-timezone/0.5.40/moment-timezone.min.js" integrity="sha512-NJfMpP34NDFAS8lJqH4FzsaD1fqoIJATgBpPjNUck9hC8kGvFhrcR8KIPnTtSinNyx8b1QPBE6NM4iux/0dHXQ==" crossorigin="anonymous" referrerpolicy="no-referrer"></script>
```

But that should be:
```
integrity="sha512-Dn6u2lfaSQfF/hd5bqRFUHVtwJWfU4osGp3OGKuUyBg39ADAh67iDwK1HWoYE/vEf2UTtZH2ZmA1fsSjA4mgWg=="
```

**Comment by MattIPv4** [comment]
I've flagged this internally again, sorry for the lack of movement.

**Comment by jpmeijers** [comment]
As mentioned in https://github.com/cdnjs/cdnjs/discussions/14124#discussioncomment-6877132 I am still being affected by this.

**Comment by dregad** [comment]
Considering that this issue has been marked as *High priority* for over 18 months and is still not resolved to date, I wonder how long it would take to fix lower priority ones 😉 

@MattIPv4 any progress with Cloudflare ? In another thread you mentioned a meeting with them in September... A resolution for this, or at least a status update would be greatly appreciated.

**Comment by dregad** [comment]
Sounds like a lost cause 😞 
I'll give it a shot, but TBH this makes me feel like maybe it's time to start looking for another CDN...

**Comment by MattIPv4** [comment]
X-posting https://github.com/cdnjs/cdnjs/issues/14210#issuecomment-1848863578:

> For the sake of complete transparency here, I've heard nothing further from folks that I have contact with at Cloudflare. As such, I've filed a formal security report through HackerOne for the fact that SRI hashes are mismatching, indicating that files on the CDN (that are meant to be immutable) have changed. Hopefully, that'll finally get some eyes on the issue. I've referenced this issue as well as #14080.

**Comment by xtuc** [comment]
https://github.com/cdnjs/tools/pull/274 fixes this issue (see examples in the PR's comment).

I went ahead and deleted the broken SRIs from the website/API. Please let me know if you find more broken SRIs, new package versions will run with the fix and produce a valid SRI.

**Comment by ioionu** [comment]
@xtuc

> Please let me know if you find more broken SRIs, new package versions will run with the fix and produce a valid SRI.

https://cdnjs.com/libraries/intro.js/7.0.1

cdnjs:

```
<script src="https://cdnjs.cloudflare.com/ajax/libs/intro.js/7.0.1/intro.min.js" integrity="sha512-S+RMGQdVq/0Rcx5sEE5TuCNTg4PEpsILZgLM2SM/X4Bz23SFrH8B3iX+HyzI+HlZB6YiAqb9M9IsNi2NRYp2Nw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script>
```

openssl sha:

```
curl -s https://cdnjs.cloudflare.com/ajax/libs/intro.js/7.0.1/intro.min.js | openssl dgst -sha512 -binary | openssl base64 -A
+xuI0HtsygKuhjjH48UnchBdvtKunhL3V/hOMZWLzFobLOP+lQdmazOn3+sa7VE6/zJKqBF5n9kOoR7S7Fwe4Q==
```

cdnjs:

```
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/intro.js/7.0.1/introjs.min.css" integrity="sha512-ZltC0pC5nP28uLJ2SBRo8EmFHA+giWPQC6nM4xzwW4hXsTaIsfSMn0qTtotOP8NNWlAohVlJtDfLRKXxA7mJ4A==" crossorigin="anonymous" referrerpolicy="no-referrer" />
```

openssl sha:

```
curl -s https://cdnjs.cloudflare.com/ajax/libs/intro.js/7.0.1/introjs.min.css | openssl dgst -sha512 -binary | openssl base64 -A
mUJxP9lPkpAyjoYrHgdHxV+GVpTKD0V9VkJ0nXvhR+hjwkCMlxM5IdoG929FDn9Dvwt5mHdW9XweOsjdKsjaNw==
```

**Comment by uob-kieren** [comment]
I think this is happening again.

`<script src="https://cdnjs.cloudflare.com/ajax/libs/intro.js/7.2.0/intro.min.js" integrity="sha512-MCM74PPZwlAz3VdPGg3cQn4Jb83jO2Q3LGika//u7H+hyk0HfyraT0u0gVf4jB44xJj3WwbGxn45L7frballmQ==" crossorigin="anonymous" referrerpolicy="no-referrer"></script>`

but:-

`curl -s https://cdnjs.cloudflare.com/ajax/libs/intro.js/7.2.0/intro.min.js | openssl dgst -sha512 -binary | openssl base64 -A
TW1wgxbyMD70qUiI6oDIoMb1fjLKF19UKAjJRVPE6lv2pfvpFIGwzhqfjXtGs3zmVREIjS1LmGgqn3GUPzQjCw==`

**Comment by drees** [comment]
> [cdnjs/tools#274](https://github.com/cdnjs/tools/pull/274) fixes this issue (see examples in the PR's comment).
> 
> I went ahead and deleted the broken SRIs from the website/API. Please let me know if you find more broken SRIs, new package versions will run with the fix and produce a valid SRI.

@xtuc [moment.js](https://cdnjs.com/libraries/moment.js) 2.30.1 is broken:
```https://cdnjs.cloudflare.com/ajax/libs/moment.js/2.30.1/moment.min.js```
cdnjs published SRI hash:
```sha512-hUhvpC5f8cgc04OZb55j0KNGh4eh7dLxd/dPSJ5VyzqDWxsayYbojWyl5Tkcgrmb/RVKCRJI1jNlRbVP4WWC4w==```
Calculated SRI hash:
```sha512-QoJS4DOhdmG8kbbHkxmB/rtPdN62cGWXAdAFWWJPvUFF1/zxcPSdAnn4HhYZSIlVoLVEJ0LesfNlusgm2bPfnA==```
